### PR TITLE
rename exceptions

### LINKED ISSUE #366: [CT-1680] Replace deprecated exception functions
`dbt-core` has deprecated functions to raise exceptions in favor of directly raising the exceptions.  Nothing should be broken (yet) and the functions still raise as expected but they also will log a deprecation warning (pending https://github.com/dbt-labs/dbt-core/pull/6413).  It would be good to resolve these deprecation messages before the 1.4 rc.

There are a *lot* of deprecated function so it's easier to search for `dbt.exceptions` and look at each callsite to determine if it's using a function or already directly raising an exception.  Then replace anywhere using functions with directly raising an exception.

We can create new exceptions in `dbt-core` to raise or the adapters could start to own their own exceptions within the adapter code base that inherit from exceptions in `dbt-core`.

For examples, see  [`dbt-core` PR](https://github.com/dbt-labs/dbt-core/pull/6347) that replaced all functions with raises.

### PR DESCRIPTION
**Pull request by emmyoop**
resolves #366 

### Description

Replace exception functions and rename exceptions.  Currently point to a branch of core for testing.

### Checklist

- [ ] I have read [the contributing guide](https://github.com/dbt-labs/dbt-snowflake/blob/main/CONTRIBUTING.md) and understand what's expected of me
- [ ] I have signed the [CLA](https://docs.getdbt.com/docs/contributor-license-agreements)
- [ ] I have run this code in development and it appears to resolve the stated issue
- [ ] This PR includes tests, or tests are not required/relevant for this PR
- [ ] I have [opened an issue to add/update docs](https://github.com/dbt-labs/docs.getdbt.com/issues/new/choose), or docs changes are not required/relevant for this PR
- [ ] I have run `changie new` to [create a changelog entry](https://github.com/dbt-labs/dbt-snowflake/blob/main/CONTRIBUTING.md#Adding-CHANGELOG-Entry)

### CODE REVIEW

File: dev-requirements.txt
Comment (VersusFacit) [inline dev-requirements.txt:3]: Could you offer some of your reasoning for why we should change this reference to a specific branch? Why not address this after merging your [dbt core-PR](https://github.com/dbt-labs/dbt-core/pull/6539) into `main`? 

I love that we were able to verify a green build on this branch, but my gut tells me we're wiser to wait to merge this until the change here no longer requires changes to dev-requirements.

**Review by VersusFacit — CHANGES_REQUESTED** [review]

File: dev-requirements.txt
Comment (emmyoop) [reply dev-requirements.txt:3]: 100% wait to merge and this will point back at main.  I just wanted to get it up for review now instead of waiting on my `dbt-core` PR to be merged in.

**Review by VersusFacit — APPROVED** [review]
LGTM. tysm!

**Pull request merged**
